`map_clone` suggests removing clone altogether

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code (https://github.com/rust-lang/rust/blob/a6ff925f8b5598a1f6d84964525baa1d4a08fd63/compiler/rustc_codegen_ssa/src/back/write.rs#L1026):

```rust
    let cgcx = CodegenContext::<B> {
        incr_comp_session_dir: sess.incr_comp_session_dir_opt().map(|r| r.clone()),
    };
```

I expected to see this happen: Not sure this lint should have fired in the first place? This is a `Ref`, not an iterator.

Instead, this happened: Clippy suggests removing the map altogether, giving a type error.

```
error[E0308]: mismatched types
    --> compiler/rustc_codegen_ssa/src/back/write.rs:1014:32
     |
1014 |         incr_comp_session_dir: sess.incr_comp_session_dir_opt(),
     |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::path::PathBuf`, found struct `std::cell::Ref`
     |
     = note: expected enum `std::option::Option<std::path::PathBuf>`
                found enum `std::option::Option<std::cell::Ref<'_, std::path::PathBuf, >>`

Original diagnostics will follow.

warning: you are needlessly cloning iterator elements
    --> compiler/rustc_codegen_ssa/src/back/write.rs:1014:64
     |
1014 |         incr_comp_session_dir: sess.incr_comp_session_dir_opt().map(|r| r.clone()),
     |                                                                ^^^^^^^^^^^^^^^^^^^ help: remove the `map` call
     |
     = note: `#[warn(clippy::map_clone)]` on by default
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone
```

### Meta

- `cargo clippy -V`: clippy 0.0.212 (ffa2e7a 2020-10-24)

--- a/clippy_lints/src/map_clone.rs
+++ b/clippy_lints/src/map_clone.rs
@@ -8,6 +8,7 @@ use rustc_hir as hir;
 use rustc_lint::{LateContext, LateLintPass};
 use rustc_middle::mir::Mutability;
 use rustc_middle::ty;
+use rustc_middle::ty::adjustment::Adjust;
 use rustc_session::{declare_lint_pass, declare_tool_lint};
 use rustc_span::symbol::Ident;
 use rustc_span::{sym, Span};
@@ -75,11 +76,14 @@ impl<'tcx> LateLintPass<'tcx> for MapClone {
                                     }
                                 }
                             },
-                            hir::ExprKind::MethodCall(ref method, _, ref obj, _) => {
-                                if ident_eq(name, &obj[0]) && method.ident.as_str() == "clone"
-                                    && match_trait_method(cx, closure_expr, &paths::CLONE_TRAIT) {
-
-                                    let obj_ty = cx.typeck_results().expr_ty(&obj[0]);
+                            hir::ExprKind::MethodCall(ref method, _, [obj], _) => if_chain! {
+                                if ident_eq(name, obj) && method.ident.name == sym::clone;
+                                if match_trait_method(cx, closure_expr, &paths::CLONE_TRAIT);
+                                // no autoderefs
+                                if !cx.typeck_results().expr_adjustments(obj).iter()
+                                    .any(|a| matches!(a.kind, Adjust::Deref(Some(..))));
+                                then {
+                                    let obj_ty = cx.typeck_results().expr_ty(obj);
                                     if let ty::Ref(_, ty, mutability) = obj_ty.kind() {
                                         if matches!(mutability, Mutability::Not) {
                                             let copy = is_copy(cx, ty);

--- a/tests/ui/map_clone.fixed
+++ b/tests/ui/map_clone.fixed
@@ -52,4 +52,11 @@ fn main() {
         let items = vec![&mut aa, &mut bb];
         let _: Vec<_> = items.into_iter().map(|x| x.clone()).collect();
     }
+
+    // Issue #6239 deref coercion and clone deref
+    {
+        use std::cell::RefCell;
+
+        let _ = Some(RefCell::new(String::new()).borrow()).map(|s| s.clone());
+    }
 }

--- a/tests/ui/map_clone.rs
+++ b/tests/ui/map_clone.rs
@@ -52,4 +52,11 @@ fn main() {
         let items = vec![&mut aa, &mut bb];
         let _: Vec<_> = items.into_iter().map(|x| x.clone()).collect();
     }
+
+    // Issue #6239 deref coercion and clone deref
+    {
+        use std::cell::RefCell;
+
+        let _ = Some(RefCell::new(String::new()).borrow()).map(|s| s.clone());
+    }
 }

